<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="3.6 Exercise IV: Single cell data | Essentials of Mathematics and Statistics" />
<meta property="og:type" content="book" />
<meta property="og:url" content="https://anasrana.github.io/module1-practical_Bham/" />

<meta property="og:description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
<meta name="github-repo" content="anasrana/module1-practical_Bham" />

<meta name="author" content="Anas A Rana" />

<meta name="date" content="2019-10-16" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>

<meta name="description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics).">

<title>3.6 Exercise IV: Single cell data | Essentials of Mathematics and Statistics</title>

<script src="libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li class="has-sub"><a href="index.html#introduction"><span class="toc-section-number">1</span> Introduction</a><ul>
<li><a href="1-1-prerequisites.html#prerequisites"><span class="toc-section-number">1.1</span> Prerequisites</a></li>
<li><a href="1-2-data-sets.html#data-sets"><span class="toc-section-number">1.2</span> Data sets</a></li>
</ul></li>
<li class="has-sub"><a href="2-practical-linear-regression.html#practical-linear-regression"><span class="toc-section-number">2</span> Practical: Linear regression</a><ul>
<li><a href="2-1-data.html#data"><span class="toc-section-number">2.1</span> Data</a></li>
<li><a href="2-2-simulating-data.html#simulating-data"><span class="toc-section-number">2.2</span> Simulating data</a></li>
<li class="has-sub"><a href="2-3-fitting-simple-linear-regression-model.html#fitting-simple-linear-regression-model"><span class="toc-section-number">2.3</span> Fitting simple linear regression model</a><ul>
<li><a href="2-3-fitting-simple-linear-regression-model.html#least-squared-estimation"><span class="toc-section-number">2.3.1</span> Least squared estimation</a></li>
<li><a href="2-3-fitting-simple-linear-regression-model.html#maximum-likelihood-estimation"><span class="toc-section-number">2.3.2</span> Maximum likelihood estimation</a></li>
</ul></li>
<li><a href="2-4-effect-of-variance.html#effect-of-variance"><span class="toc-section-number">2.4</span> Effect of variance</a></li>
<li class="has-sub"><a href="2-5-exercise.html#exercise"><span class="toc-section-number">2.5</span> Exercise</a><ul>
<li><a href="2-5-exercise.html#part-i"><span class="toc-section-number">2.5.1</span> Part I</a></li>
<li><a href="2-5-exercise.html#part-ii"><span class="toc-section-number">2.5.2</span> Part II</a></li>
<li><a href="2-5-exercise.html#part-iii"><span class="toc-section-number">2.5.3</span> Part III</a></li>
</ul></li>
</ul></li>
<li class="has-sub"><a href="model-answers-linear-regression.html#model-answers-linear-regression">Model answers: Linear regression</a><ul>
<li><a href="2-6-exercise-i.html#exercise-i"><span class="toc-section-number">2.6</span> Exercise I</a></li>
<li class="has-sub"><a href="2-7-exercise-ii.html#exercise-ii"><span class="toc-section-number">2.7</span> Exercise II</a><ul>
<li><a href="2-7-exercise-ii.html#comparison-of-data"><span class="toc-section-number">2.7.1</span> Comparison of data</a></li>
</ul></li>
<li><a href="2-8-exercise-ii-1.html#exercise-ii-1"><span class="toc-section-number">2.8</span> Exercise II</a></li>
</ul></li>
<li class="has-sub"><a href="3-practical-principal-component-analysis.html#practical-principal-component-analysis"><span class="toc-section-number">3</span> Practical: Principal component analysis</a><ul>
<li><a href="3-1-data-1.html#data-1"><span class="toc-section-number">3.1</span> Data</a></li>
<li><a href="3-2-introduction-1.html#introduction-1"><span class="toc-section-number">3.2</span> Introduction</a></li>
<li><a href="3-3-exercise-i-1.html#exercise-i-1"><span class="toc-section-number">3.3</span> Exercise I</a></li>
<li><a href="3-4-exercise-ii-2.html#exercise-ii-2"><span class="toc-section-number">3.4</span> Exercise II</a></li>
<li><a href="3-5-exercise-iii.html#exercise-iii"><span class="toc-section-number">3.5</span> Exercise III</a></li>
<li><a href="3-6-exercise-iv-single-cell-data.html#exercise-iv-single-cell-data"><span class="toc-section-number">3.6</span> Exercise IV: Single cell data</a></li>
</ul></li>
<li class="has-sub"><a href="4-practical-multiple-regression.html#practical-multiple-regression"><span class="toc-section-number">4</span> Practical: Multiple regression</a><ul>
<li><a href="4-1-multiple-regression.html#multiple-regression"><span class="toc-section-number">4.1</span> Multiple regression</a></li>
<li><a href="4-2-categorical-covariates.html#categorical-covariates"><span class="toc-section-number">4.2</span> Categorical covariates</a></li>
<li><a href="4-3-residuals.html#residuals"><span class="toc-section-number">4.3</span> Residuals</a></li>
<li><a href="4-4-gradient-descent-algorithm.html#gradient-descent-algorithm"><span class="toc-section-number">4.4</span> Gradient descent algorithm (+)</a></li>
</ul></li>
<li class="has-sub"><a href="5-practical-generalised-linear-models.html#practical-generalised-linear-models"><span class="toc-section-number">5</span> Practical: Generalised linear models</a><ul>
<li><a href="5-1-data-2.html#data-2"><span class="toc-section-number">5.1</span> Data</a></li>
<li><a href="5-2-detecting-snp-associations.html#detecting-snp-associations"><span class="toc-section-number">5.2</span> Detecting SNP associations</a></li>
<li><a href="5-3-gwas-and-logistic-regression.html#gwas-and-logistic-regression"><span class="toc-section-number">5.3</span> GWAS and logistic regression</a></li>
<li class="has-sub"><a href="5-4-negative-binomial-and-poisson-regression.html#negative-binomial-and-poisson-regression"><span class="toc-section-number">5.4</span> Negative binomial and Poisson regression</a><ul>
<li><a href="5-4-negative-binomial-and-poisson-regression.html#count-based-glms"><span class="toc-section-number">5.4.1</span> Count-based GLMs</a></li>
<li><a href="5-4-negative-binomial-and-poisson-regression.html#fitting-a-glm"><span class="toc-section-number">5.4.2</span> Fitting a GLM</a></li>
<li><a href="5-4-negative-binomial-and-poisson-regression.html#comparing-nested-models"><span class="toc-section-number">5.4.3</span> Comparing nested models</a></li>
</ul></li>
<li><a href="5-5-negative-binomial-vs-poisson-glms.html#negative-binomial-vs-poisson-glms"><span class="toc-section-number">5.5</span> Negative-Binomial vs Poisson GLMs</a></li>
<li><a href="5-6-further-understanding-the-model-optional.html#further-understanding-the-model-optional"><span class="toc-section-number">5.6</span> Further understanding the model (<strong>OPTIONAL</strong>)</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="exercise-iv-single-cell-data" class="section level2">
<h2><span class="header-section-number">3.6</span> Exercise IV: Single cell data</h2>
<p>We can now try to apply what we learned above on a more realistic datasets. You can download the data either on <em>canvas</em> or using these links <a href="https://raw.github.com/anasrana/module1-practical_Bham/master/data/Pollen2014.txt"><code>Pollen2014.txt</code></a> and <a href="https://raw.github.com/anasrana/module1-practical_Bham/master/data/SupplementaryLabels.txt"><code>SupplementaryLabels.txt</code></a>. Her we will be dealing with single cell RNA-Seq (scRNA-Seq) data, which consist of <span class="math inline">\(300\)</span> single cells measured across <span class="math inline">\(8686\)</span> genes.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="3-6-exercise-iv-single-cell-data.html#cb60-1"></a>pollen_df &lt;-<span class="kw">read.table</span>(<span class="st">&quot;Pollen2014.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&#39;,&#39;</span>, <span class="dt">header =</span> T,<span class="dt">row.names=</span><span class="dv">1</span>)</span>
<span id="cb60-2"><a href="3-6-exercise-iv-single-cell-data.html#cb60-2"></a></span>
<span id="cb60-3"><a href="3-6-exercise-iv-single-cell-data.html#cb60-3"></a>label_df &lt;-<span class="kw">read.table</span>(<span class="st">&quot;SupplementaryLabels.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&#39;,&#39;</span>, <span class="dt">header =</span> T)</span>
<span id="cb60-4"><a href="3-6-exercise-iv-single-cell-data.html#cb60-4"></a></span>
<span id="cb60-5"><a href="3-6-exercise-iv-single-cell-data.html#cb60-5"></a>pollen_df[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</span></code></pre></div>
<pre><code>##          Cell_2338_1 Cell_2338_10 Cell_2338_11 Cell_2338_12 Cell_2338_13
## MTND2P28          78          559          811          705          384
## MTATP6P1        2053         1958         4922         4409         2610
## NOC2L              1          125          126            0          487
## ISG15           2953         4938          580          523         2609
## CPSF3L             2           42           19            0           37
## MXRA8              0            0            0            0            0
## AURKAIP1         302          132           64          492           11
## CCNL2              0          235            0           84           13
## MRPL20           330          477          288          222           44
## SSU72            604          869         2046          158          530
##          Cell_2338_14
## MTND2P28          447
## MTATP6P1         3709
## NOC2L              66
## ISG15               1
## CPSF3L             12
## MXRA8               0
## AURKAIP1          182
## CCNL2              11
## MRPL20            282
## SSU72             272</code></pre>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="3-6-exercise-iv-single-cell-data.html#cb62-1"></a><span class="kw">dim</span>(pollen_df)</span></code></pre></div>
<pre><code>## [1] 8686  300</code></pre>
<p>Measurements of scRNA-Seq data are integer counts, this data does not have good properties so we perform a transformation on the data. The most commonly used transformation on RNA-Seq count data is <span class="math inline">\(\log_2\)</span>. We will also transpose the data matrix to rows representing cells and columns representing genes. This is the data we can use to perform PCA.</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="3-6-exercise-iv-single-cell-data.html#cb64-1"></a><span class="co"># scRNA-Seq data transformation</span></span>
<span id="cb64-2"><a href="3-6-exercise-iv-single-cell-data.html#cb64-2"></a>pollen_mat &lt;-<span class="st"> </span><span class="kw">log2</span>(<span class="kw">as.matrix</span>(pollen_df) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)</span>
<span id="cb64-3"><a href="3-6-exercise-iv-single-cell-data.html#cb64-3"></a><span class="co"># transpose the data</span></span>
<span id="cb64-4"><a href="3-6-exercise-iv-single-cell-data.html#cb64-4"></a>pollen_mat &lt;-<span class="st"> </span><span class="kw">t</span>(pollen_mat)</span></code></pre></div>
<p>We will now use information that we read into the <code>label_df</code> variable to rename cells.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="3-6-exercise-iv-single-cell-data.html#cb65-1"></a><span class="co"># Check which columns we have available</span></span>
<span id="cb65-2"><a href="3-6-exercise-iv-single-cell-data.html#cb65-2"></a><span class="kw">colnames</span>(label_df)</span></code></pre></div>
<pre><code>## [1] &quot;Cell_Identifier&quot;        &quot;Population&quot;            
## [3] &quot;Cell_names&quot;             &quot;TrueLabel_CellLevel&quot;   
## [5] &quot;Tissue_name&quot;            &quot;TrueLabel_TissuelLevel&quot;</code></pre>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="3-6-exercise-iv-single-cell-data.html#cb67-1"></a><span class="co"># rename rows</span></span>
<span id="cb67-2"><a href="3-6-exercise-iv-single-cell-data.html#cb67-2"></a><span class="kw">rownames</span>(pollen_mat) &lt;-<span class="st"> </span>label_df<span class="op">$</span>Cell_names</span></code></pre></div>
<p>Next we perform PCA on the data and extract the proportion of variance explained by each component.</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="3-6-exercise-iv-single-cell-data.html#cb68-1"></a>sc_pca &lt;-<span class="st"> </span><span class="kw">prcomp</span>(pollen_mat)</span>
<span id="cb68-2"><a href="3-6-exercise-iv-single-cell-data.html#cb68-2"></a></span>
<span id="cb68-3"><a href="3-6-exercise-iv-single-cell-data.html#cb68-3"></a><span class="co"># variance is the square of the standard deviation</span></span>
<span id="cb68-4"><a href="3-6-exercise-iv-single-cell-data.html#cb68-4"></a>pr_var &lt;-<span class="st"> </span>sc_pca<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span></span>
<span id="cb68-5"><a href="3-6-exercise-iv-single-cell-data.html#cb68-5"></a></span>
<span id="cb68-6"><a href="3-6-exercise-iv-single-cell-data.html#cb68-6"></a><span class="co"># compute the variance explained by each principal component</span></span>
<span id="cb68-7"><a href="3-6-exercise-iv-single-cell-data.html#cb68-7"></a>prop_var_exp &lt;-<span class="st"> </span>pr_var <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(pr_var)</span></code></pre></div>
<p>Think about the calculation and what exactly it means. We can visualise this</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="3-6-exercise-iv-single-cell-data.html#cb69-1"></a>var_exp &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">variance =</span> prop_var_exp, <span class="dt">pc =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(prop_var_exp))</span>
<span id="cb69-2"><a href="3-6-exercise-iv-single-cell-data.html#cb69-2"></a></span>
<span id="cb69-3"><a href="3-6-exercise-iv-single-cell-data.html#cb69-3"></a><span class="kw">ggplot</span>(var_exp[<span class="dv">1</span><span class="op">:</span><span class="dv">30</span>, ], <span class="kw">aes</span>(<span class="dt">x =</span> pc, <span class="dt">y =</span> variance)) <span class="op">+</span></span>
<span id="cb69-4"><a href="3-6-exercise-iv-single-cell-data.html#cb69-4"></a><span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">&quot;identity&quot;</span>) <span class="op">+</span></span>
<span id="cb69-5"><a href="3-6-exercise-iv-single-cell-data.html#cb69-5"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Principal Component&quot;</span>,</span>
<span id="cb69-6"><a href="3-6-exercise-iv-single-cell-data.html#cb69-6"></a>         <span class="dt">y  =</span> <span class="st">&quot;Variance explained&quot;</span>)</span></code></pre></div>
<p><img src="Practicals-module_1_files/figure-html/pca-scree-1.png" width="3000" /></p>
<p>We see that the first few principal components explain significant variance, but after about the PC10, there is very limited contribution. Next we will plot the data using the first two Principal components as before.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="3-6-exercise-iv-single-cell-data.html#cb70-1"></a>sc_pca_df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(sc_pca<span class="op">$</span>x, <span class="dt">cell =</span> <span class="kw">rownames</span>(sc_pca<span class="op">$</span>x),</span>
<span id="cb70-2"><a href="3-6-exercise-iv-single-cell-data.html#cb70-2"></a>                        <span class="dt">var_exp =</span> prop_var_exp)</span>
<span id="cb70-3"><a href="3-6-exercise-iv-single-cell-data.html#cb70-3"></a></span>
<span id="cb70-4"><a href="3-6-exercise-iv-single-cell-data.html#cb70-4"></a><span class="kw">ggplot</span>(sc_pca_df, <span class="kw">aes</span>(<span class="dt">x =</span> PC1, <span class="dt">y =</span> PC2, <span class="dt">col =</span> cell)) <span class="op">+</span></span>
<span id="cb70-5"><a href="3-6-exercise-iv-single-cell-data.html#cb70-5"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span></span>
<span id="cb70-6"><a href="3-6-exercise-iv-single-cell-data.html#cb70-6"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&quot;bottom&quot;</span>)</span></code></pre></div>
<p><img src="Practicals-module_1_files/figure-html/sc-pca_plot-1.png" width="3000" /></p>
<p>Why is it not useful to create biplot for this example?</p>

</div>
<!-- </div> -->
<p style="text-align: center;">
<a href="3-5-exercise-iii.html"><button class="btn btn-default">Previous</button></a>
<a href="4-practical-multiple-regression.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
