<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="5.2 Detecting SNP associations | Essentials of Mathematics and Statistics" />
<meta property="og:type" content="book" />
<meta property="og:url" content="https://anasrana.github.io/module1-practical_Bham/" />

<meta property="og:description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
<meta name="github-repo" content="anasrana/module1-practical_Bham" />

<meta name="author" content="Anas A Rana" />

<meta name="date" content="2019-10-16" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>

<meta name="description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics).">

<title>5.2 Detecting SNP associations | Essentials of Mathematics and Statistics</title>

<script src="libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="libs/navigation-1.1/tabsets.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
</style>
</head>

<body>

<div class="container-fluid main-container">


<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul>
<li class="has-sub"><a href="index.html#introduction"><span class="toc-section-number">1</span> Introduction</a><ul>
<li><a href="1-1-prerequisites.html#prerequisites"><span class="toc-section-number">1.1</span> Prerequisites</a></li>
<li><a href="1-2-data-sets.html#data-sets"><span class="toc-section-number">1.2</span> Data sets</a></li>
</ul></li>
<li class="has-sub"><a href="2-practical-linear-regression.html#practical-linear-regression"><span class="toc-section-number">2</span> Practical: Linear regression</a><ul>
<li><a href="2-1-data.html#data"><span class="toc-section-number">2.1</span> Data</a></li>
<li><a href="2-2-simulating-data.html#simulating-data"><span class="toc-section-number">2.2</span> Simulating data</a></li>
<li class="has-sub"><a href="2-3-fitting-simple-linear-regression-model.html#fitting-simple-linear-regression-model"><span class="toc-section-number">2.3</span> Fitting simple linear regression model</a><ul>
<li><a href="2-3-fitting-simple-linear-regression-model.html#least-squared-estimation"><span class="toc-section-number">2.3.1</span> Least squared estimation</a></li>
<li><a href="2-3-fitting-simple-linear-regression-model.html#maximum-likelihood-estimation"><span class="toc-section-number">2.3.2</span> Maximum likelihood estimation</a></li>
</ul></li>
<li><a href="2-4-effect-of-variance.html#effect-of-variance"><span class="toc-section-number">2.4</span> Effect of variance</a></li>
<li class="has-sub"><a href="2-5-exercise.html#exercise"><span class="toc-section-number">2.5</span> Exercise</a><ul>
<li><a href="2-5-exercise.html#part-i"><span class="toc-section-number">2.5.1</span> Part I</a></li>
<li><a href="2-5-exercise.html#part-ii"><span class="toc-section-number">2.5.2</span> Part II</a></li>
<li><a href="2-5-exercise.html#part-iii"><span class="toc-section-number">2.5.3</span> Part III</a></li>
</ul></li>
</ul></li>
<li class="has-sub"><a href="model-answers-linear-regression.html#model-answers-linear-regression">Model answers: Linear regression</a><ul>
<li><a href="2-6-exercise-i.html#exercise-i"><span class="toc-section-number">2.6</span> Exercise I</a></li>
<li class="has-sub"><a href="2-7-exercise-ii.html#exercise-ii"><span class="toc-section-number">2.7</span> Exercise II</a><ul>
<li><a href="2-7-exercise-ii.html#comparison-of-data"><span class="toc-section-number">2.7.1</span> Comparison of data</a></li>
</ul></li>
<li><a href="2-8-exercise-ii-1.html#exercise-ii-1"><span class="toc-section-number">2.8</span> Exercise II</a></li>
</ul></li>
<li class="has-sub"><a href="3-practical-principal-component-analysis.html#practical-principal-component-analysis"><span class="toc-section-number">3</span> Practical: Principal component analysis</a><ul>
<li><a href="3-1-data-1.html#data-1"><span class="toc-section-number">3.1</span> Data</a></li>
<li><a href="3-2-introduction-1.html#introduction-1"><span class="toc-section-number">3.2</span> Introduction</a></li>
<li><a href="3-3-exercise-i-1.html#exercise-i-1"><span class="toc-section-number">3.3</span> Exercise I</a></li>
<li><a href="3-4-exercise-ii-2.html#exercise-ii-2"><span class="toc-section-number">3.4</span> Exercise II</a></li>
<li><a href="3-5-exercise-iii.html#exercise-iii"><span class="toc-section-number">3.5</span> Exercise III</a></li>
<li><a href="3-6-exercise-iv-single-cell-data.html#exercise-iv-single-cell-data"><span class="toc-section-number">3.6</span> Exercise IV: Single cell data</a></li>
</ul></li>
<li class="has-sub"><a href="4-practical-multiple-regression.html#practical-multiple-regression"><span class="toc-section-number">4</span> Practical: Multiple regression</a><ul>
<li><a href="4-1-multiple-regression.html#multiple-regression"><span class="toc-section-number">4.1</span> Multiple regression</a></li>
<li><a href="4-2-categorical-covariates.html#categorical-covariates"><span class="toc-section-number">4.2</span> Categorical covariates</a></li>
<li><a href="4-3-residuals.html#residuals"><span class="toc-section-number">4.3</span> Residuals</a></li>
<li><a href="4-4-gradient-descent-algorithm.html#gradient-descent-algorithm"><span class="toc-section-number">4.4</span> Gradient descent algorithm (+)</a></li>
</ul></li>
<li class="has-sub"><a href="5-practical-generalised-linear-models.html#practical-generalised-linear-models"><span class="toc-section-number">5</span> Practical: Generalised linear models</a><ul>
<li><a href="5-1-data-2.html#data-2"><span class="toc-section-number">5.1</span> Data</a></li>
<li><a href="5-2-detecting-snp-associations.html#detecting-snp-associations"><span class="toc-section-number">5.2</span> Detecting SNP associations</a></li>
<li><a href="5-3-gwas-and-logistic-regression.html#gwas-and-logistic-regression"><span class="toc-section-number">5.3</span> GWAS and logistic regression</a></li>
<li class="has-sub"><a href="5-4-negative-binomial-and-poisson-regression.html#negative-binomial-and-poisson-regression"><span class="toc-section-number">5.4</span> Negative binomial and Poisson regression</a><ul>
<li><a href="5-4-negative-binomial-and-poisson-regression.html#count-based-glms"><span class="toc-section-number">5.4.1</span> Count-based GLMs</a></li>
<li><a href="5-4-negative-binomial-and-poisson-regression.html#fitting-a-glm"><span class="toc-section-number">5.4.2</span> Fitting a GLM</a></li>
<li><a href="5-4-negative-binomial-and-poisson-regression.html#comparing-nested-models"><span class="toc-section-number">5.4.3</span> Comparing nested models</a></li>
</ul></li>
<li><a href="5-5-negative-binomial-vs-poisson-glms.html#negative-binomial-vs-poisson-glms"><span class="toc-section-number">5.5</span> Negative-Binomial vs Poisson GLMs</a></li>
<li><a href="5-6-further-understanding-the-model-optional.html#further-understanding-the-model-optional"><span class="toc-section-number">5.6</span> Further understanding the model (<strong>OPTIONAL</strong>)</a></li>
</ul></li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<div id="detecting-snp-associations" class="section level2">
<h2><span class="header-section-number">5.2</span> Detecting SNP associations</h2>
<p>We have seen in lectures that we can do statistical tests for this type of contingency table using Chi Squared Tests. Let’s load example data set and and prepare</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="5-2-detecting-snp-associations.html#cb97-1"></a><span class="kw">library</span>(ggplot2) <span class="co"># for plots later</span></span>
<span id="cb97-2"><a href="5-2-detecting-snp-associations.html#cb97-2"></a><span class="kw">load</span>(<span class="st">&quot;gwas-cc-ex1.Rdata&quot;</span>)</span>
<span id="cb97-3"><a href="5-2-detecting-snp-associations.html#cb97-3"></a></span>
<span id="cb97-4"><a href="5-2-detecting-snp-associations.html#cb97-4"></a><span class="co"># how many individuals are there</span></span>
<span id="cb97-5"><a href="5-2-detecting-snp-associations.html#cb97-5"></a>n &lt;-<span class="st"> </span><span class="kw">length</span>(y)</span>
<span id="cb97-6"><a href="5-2-detecting-snp-associations.html#cb97-6"></a><span class="co"># How many SNPs do we have data for</span></span>
<span id="cb97-7"><a href="5-2-detecting-snp-associations.html#cb97-7"></a>p &lt;-<span class="st"> </span><span class="kw">nrow</span>(X)</span>
<span id="cb97-8"><a href="5-2-detecting-snp-associations.html#cb97-8"></a></span>
<span id="cb97-9"><a href="5-2-detecting-snp-associations.html#cb97-9"></a><span class="co"># samples that are controls are encoded as 0 in y</span></span>
<span id="cb97-10"><a href="5-2-detecting-snp-associations.html#cb97-10"></a>control &lt;-<span class="st"> </span><span class="kw">which</span>(y <span class="op">==</span><span class="st"> </span><span class="dv">0</span>)</span>
<span id="cb97-11"><a href="5-2-detecting-snp-associations.html#cb97-11"></a><span class="co"># disease cases are encoded as 1 in y</span></span>
<span id="cb97-12"><a href="5-2-detecting-snp-associations.html#cb97-12"></a>cases &lt;-<span class="st"> </span><span class="kw">which</span>(y <span class="op">==</span><span class="st"> </span><span class="dv">1</span>)</span></code></pre></div>
<p>Now we need to write a loop that scans through all, <span class="math inline">\(p\)</span>, SNPs:</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="5-2-detecting-snp-associations.html#cb98-1"></a><span class="co"># create a vector where p-values will be stored</span></span>
<span id="cb98-2"><a href="5-2-detecting-snp-associations.html#cb98-2"></a>p_vals &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, p)</span>
<span id="cb98-3"><a href="5-2-detecting-snp-associations.html#cb98-3"></a></span>
<span id="cb98-4"><a href="5-2-detecting-snp-associations.html#cb98-4"></a><span class="co"># Loop over SNPs</span></span>
<span id="cb98-5"><a href="5-2-detecting-snp-associations.html#cb98-5"></a><span class="cf">for</span> (i_p <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>p) {</span>
<span id="cb98-6"><a href="5-2-detecting-snp-associations.html#cb98-6"></a>    <span class="co"># 1. obtain genotype counts</span></span>
<span id="cb98-7"><a href="5-2-detecting-snp-associations.html#cb98-7"></a>    counts &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">3</span>)</span>
<span id="cb98-8"><a href="5-2-detecting-snp-associations.html#cb98-8"></a>    counts[<span class="dv">1</span>, ] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>),</span>
<span id="cb98-9"><a href="5-2-detecting-snp-associations.html#cb98-9"></a>                     <span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>),</span>
<span id="cb98-10"><a href="5-2-detecting-snp-associations.html#cb98-10"></a>                     <span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">2</span>))</span>
<span id="cb98-11"><a href="5-2-detecting-snp-associations.html#cb98-11"></a></span>
<span id="cb98-12"><a href="5-2-detecting-snp-associations.html#cb98-12"></a>    counts[<span class="dv">2</span>, ] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>),</span>
<span id="cb98-13"><a href="5-2-detecting-snp-associations.html#cb98-13"></a>                     <span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>),</span>
<span id="cb98-14"><a href="5-2-detecting-snp-associations.html#cb98-14"></a>                     <span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">2</span>))</span>
<span id="cb98-15"><a href="5-2-detecting-snp-associations.html#cb98-15"></a></span>
<span id="cb98-16"><a href="5-2-detecting-snp-associations.html#cb98-16"></a>    <span class="co"># 2. expected probability of AA</span></span>
<span id="cb98-17"><a href="5-2-detecting-snp-associations.html#cb98-17"></a>    <span class="co"># (assuming no dependence on case/control status)</span></span>
<span id="cb98-18"><a href="5-2-detecting-snp-associations.html#cb98-18"></a>    expected_pr_AA &lt;-<span class="st"> </span><span class="kw">sum</span>(counts[, <span class="dv">1</span>]) <span class="op">/</span><span class="st"> </span>n</span>
<span id="cb98-19"><a href="5-2-detecting-snp-associations.html#cb98-19"></a>    <span class="co"># expected probability of Aa</span></span>
<span id="cb98-20"><a href="5-2-detecting-snp-associations.html#cb98-20"></a>    expected_pr_Aa &lt;-<span class="st"> </span><span class="kw">sum</span>(counts[, <span class="dv">2</span>]) <span class="op">/</span><span class="st"> </span>n</span>
<span id="cb98-21"><a href="5-2-detecting-snp-associations.html#cb98-21"></a>    <span class="co"># expected probability of aa</span></span>
<span id="cb98-22"><a href="5-2-detecting-snp-associations.html#cb98-22"></a>    expected_pr_aa &lt;-<span class="st"> </span><span class="kw">sum</span>(counts[, <span class="dv">3</span>]) <span class="op">/</span><span class="st"> </span>n</span>
<span id="cb98-23"><a href="5-2-detecting-snp-associations.html#cb98-23"></a></span>
<span id="cb98-24"><a href="5-2-detecting-snp-associations.html#cb98-24"></a>    expected_probs &lt;-<span class="st"> </span><span class="kw">c</span>(expected_pr_AA, expected_pr_Aa, expected_pr_aa )</span>
<span id="cb98-25"><a href="5-2-detecting-snp-associations.html#cb98-25"></a></span>
<span id="cb98-26"><a href="5-2-detecting-snp-associations.html#cb98-26"></a>    <span class="co"># 3. do my chi-squared test</span></span>
<span id="cb98-27"><a href="5-2-detecting-snp-associations.html#cb98-27"></a>    out &lt;-<span class="st"> </span><span class="kw">chisq.test</span>(counts, <span class="dt">p =</span> expected_probs)</span>
<span id="cb98-28"><a href="5-2-detecting-snp-associations.html#cb98-28"></a>    <span class="co"># extract p value of test and store</span></span>
<span id="cb98-29"><a href="5-2-detecting-snp-associations.html#cb98-29"></a>    p_vals[i_p] &lt;-<span class="st"> </span>out<span class="op">$</span>p.value</span>
<span id="cb98-30"><a href="5-2-detecting-snp-associations.html#cb98-30"></a>}</span></code></pre></div>
<p>We went through each SNP (rows in matrix <code>X</code>), extracted the counts of each genotype (marked <code>1.</code> in code) for cases and controls, then we compute expected probability (marked <code>2.</code> in code). Finally, we perform a chi-squared contingency table test comparing those observed counts to expected probabilities assuming that genotype is not related to disease status (marked <code>3.</code> in code).</p>
<pre><code>## You will likely need to install these fonts on your system as well.
## 
## You can find them in [/Users/ranaaaa/Library/R/3.6/library/hrbrthemes/fonts/plex-sans]</code></pre>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="5-2-detecting-snp-associations.html#cb100-1"></a>p_val_df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">p_val =</span> p_vals, <span class="dt">idx =</span> <span class="dv">1</span><span class="op">:</span>p)</span>
<span id="cb100-2"><a href="5-2-detecting-snp-associations.html#cb100-2"></a></span>
<span id="cb100-3"><a href="5-2-detecting-snp-associations.html#cb100-3"></a><span class="kw">ggplot</span>(p_val_df, <span class="kw">aes</span>(<span class="dt">x =</span> idx, <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(p_val))) <span class="op">+</span></span>
<span id="cb100-4"><a href="5-2-detecting-snp-associations.html#cb100-4"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="fl">2.5</span>, <span class="dt">col =</span> <span class="st">&quot;dodgerblue1&quot;</span>)</span></code></pre></div>
<p><img src="Practicals-module_1_files/figure-html/plot-1.png" width="3000" /></p>
<p>This plot is knows as a <em>Manhattan</em> plot. One SNP (<code>i_p = 250</code>) will pop out as being highly associated with the disease process. Look at the genotype counts (or MAF) for this SNP in the cases and controls to see for yourself that there is large difference in the distribution of genotypes (or MAF).</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="5-2-detecting-snp-associations.html#cb101-1"></a>i_p &lt;-<span class="st"> </span><span class="dv">250</span></span>
<span id="cb101-2"><a href="5-2-detecting-snp-associations.html#cb101-2"></a>counts_v &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>), <span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>),</span>
<span id="cb101-3"><a href="5-2-detecting-snp-associations.html#cb101-3"></a>              <span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">2</span>), <span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>),</span>
<span id="cb101-4"><a href="5-2-detecting-snp-associations.html#cb101-4"></a>              <span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>), <span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">2</span>))</span>
<span id="cb101-5"><a href="5-2-detecting-snp-associations.html#cb101-5"></a></span>
<span id="cb101-6"><a href="5-2-detecting-snp-associations.html#cb101-6"></a>snp_procs &lt;-<span class="st"> </span><span class="kw">data.frame</span>(counts_v, <span class="dt">type =</span> <span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;control&quot;</span>, <span class="st">&quot;cases&quot;</span>), <span class="dt">each =</span> <span class="dv">3</span>),</span>
<span id="cb101-7"><a href="5-2-detecting-snp-associations.html#cb101-7"></a>           <span class="dt">genotype =</span> <span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;AA&quot;</span>, <span class="st">&quot;Aa&quot;</span>, <span class="st">&quot;aa&quot;</span>), <span class="dv">2</span>))</span>
<span id="cb101-8"><a href="5-2-detecting-snp-associations.html#cb101-8"></a></span>
<span id="cb101-9"><a href="5-2-detecting-snp-associations.html#cb101-9"></a><span class="kw">ggplot</span>(snp_procs, <span class="kw">aes</span>(<span class="dt">x =</span> genotype, <span class="dt">y =</span> counts_v, <span class="dt">fill =</span> type)) <span class="op">+</span></span>
<span id="cb101-10"><a href="5-2-detecting-snp-associations.html#cb101-10"></a><span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">&quot;identity&quot;</span>, <span class="dt">position =</span> <span class="st">&quot;dodge&quot;</span>)</span></code></pre></div>
<p><img src="Practicals-module_1_files/figure-html/snp-1.png" width="2400" /></p>
</div>
<p style="text-align: center;">
<a href="5-1-data-2.html"><button class="btn btn-default">Previous</button></a>
<a href="5-3-gwas-and-logistic-regression.html"><button class="btn btn-default">Next</button></a>
</p>
</div>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

</body>
</html>
