<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5.2 Detecting SNP associations | Essentials of Mathematics and Statistics</title>
  <meta name="description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
  <meta name="generator" content="bookdown 0.14 and GitBook 2.6.7" />

  <meta property="og:title" content="5.2 Detecting SNP associations | Essentials of Mathematics and Statistics" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://anasrana.github.io/module1-practical_Bham/" />
  
  <meta property="og:description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
  <meta name="github-repo" content="anasrana/module1-practical_Bham" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5.2 Detecting SNP associations | Essentials of Mathematics and Statistics" />
  
  <meta name="twitter:description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
  

<meta name="author" content="Anas A Rana" />


<meta name="date" content="2019-10-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-2.html"/>
<link rel="next" href="gwas-and-logistic-regression.html"/>
<script src="assets/jquery-2.2.3/jquery.min.js"></script>
<link href="assets/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Essentials of Mathematics and Statistics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="prerequisites.html"><a href="prerequisites.html"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="data-sets.html"><a href="data-sets.html"><i class="fa fa-check"></i><b>1.2</b> Data sets</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="practical-linear-regression.html"><a href="practical-linear-regression.html"><i class="fa fa-check"></i><b>2</b> Practical: Linear regression</a><ul>
<li class="chapter" data-level="2.1" data-path="data.html"><a href="data.html"><i class="fa fa-check"></i><b>2.1</b> Data</a></li>
<li class="chapter" data-level="2.2" data-path="simulating-data.html"><a href="simulating-data.html"><i class="fa fa-check"></i><b>2.2</b> Simulating data</a></li>
<li class="chapter" data-level="2.3" data-path="fitting-simple-linear-regression-model.html"><a href="fitting-simple-linear-regression-model.html"><i class="fa fa-check"></i><b>2.3</b> Fitting simple linear regression model</a><ul>
<li class="chapter" data-level="2.3.1" data-path="fitting-simple-linear-regression-model.html"><a href="fitting-simple-linear-regression-model.html#least-squared-estimation"><i class="fa fa-check"></i><b>2.3.1</b> Least squared estimation</a></li>
<li class="chapter" data-level="2.3.2" data-path="fitting-simple-linear-regression-model.html"><a href="fitting-simple-linear-regression-model.html#maximum-likelihood-estimation"><i class="fa fa-check"></i><b>2.3.2</b> Maximum likelihood estimation</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="effect-of-variance.html"><a href="effect-of-variance.html"><i class="fa fa-check"></i><b>2.4</b> Effect of variance</a></li>
<li class="chapter" data-level="2.5" data-path="exercise.html"><a href="exercise.html"><i class="fa fa-check"></i><b>2.5</b> Exercise</a><ul>
<li class="chapter" data-level="2.5.1" data-path="exercise.html"><a href="exercise.html#part-i"><i class="fa fa-check"></i><b>2.5.1</b> Part I</a></li>
<li class="chapter" data-level="2.5.2" data-path="exercise.html"><a href="exercise.html#part-ii"><i class="fa fa-check"></i><b>2.5.2</b> Part II</a></li>
<li class="chapter" data-level="2.5.3" data-path="exercise.html"><a href="exercise.html#part-iii"><i class="fa fa-check"></i><b>2.5.3</b> Part III</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="model-answers-linear-regression.html"><a href="model-answers-linear-regression.html"><i class="fa fa-check"></i>Model answers: Linear regression</a><ul>
<li class="chapter" data-level="2.6" data-path="exercise-i.html"><a href="exercise-i.html"><i class="fa fa-check"></i><b>2.6</b> Exercise I</a></li>
<li class="chapter" data-level="2.7" data-path="exercise-ii.html"><a href="exercise-ii.html"><i class="fa fa-check"></i><b>2.7</b> Exercise II</a><ul>
<li class="chapter" data-level="2.7.1" data-path="exercise-ii.html"><a href="exercise-ii.html#comparison-of-data"><i class="fa fa-check"></i><b>2.7.1</b> Comparison of data</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="exercise-ii-1.html"><a href="exercise-ii-1.html"><i class="fa fa-check"></i><b>2.8</b> Exercise II</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="practical-principal-component-analysis.html"><a href="practical-principal-component-analysis.html"><i class="fa fa-check"></i><b>3</b> Practical: Principal component analysis</a><ul>
<li class="chapter" data-level="3.1" data-path="data-1.html"><a href="data-1.html"><i class="fa fa-check"></i><b>3.1</b> Data</a></li>
<li class="chapter" data-level="3.2" data-path="introduction-1.html"><a href="introduction-1.html"><i class="fa fa-check"></i><b>3.2</b> Introduction</a></li>
<li class="chapter" data-level="3.3" data-path="exercise-i-1.html"><a href="exercise-i-1.html"><i class="fa fa-check"></i><b>3.3</b> Exercise I</a></li>
<li class="chapter" data-level="3.4" data-path="exercise-ii-2.html"><a href="exercise-ii-2.html"><i class="fa fa-check"></i><b>3.4</b> Exercise II</a></li>
<li class="chapter" data-level="3.5" data-path="exercise-iii.html"><a href="exercise-iii.html"><i class="fa fa-check"></i><b>3.5</b> Exercise III</a></li>
<li class="chapter" data-level="3.6" data-path="exercise-iv-single-cell-data.html"><a href="exercise-iv-single-cell-data.html"><i class="fa fa-check"></i><b>3.6</b> Exercise IV: Single cell data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="practical-multivariate-regression.html"><a href="practical-multivariate-regression.html"><i class="fa fa-check"></i><b>4</b> Practical: Multivariate regression</a></li>
<li class="chapter" data-level="5" data-path="practical-generalised-linear-models.html"><a href="practical-generalised-linear-models.html"><i class="fa fa-check"></i><b>5</b> Practical: Generalised linear models</a><ul>
<li class="chapter" data-level="5.1" data-path="data-2.html"><a href="data-2.html"><i class="fa fa-check"></i><b>5.1</b> Data</a></li>
<li class="chapter" data-level="5.2" data-path="detecting-snp-associations.html"><a href="detecting-snp-associations.html"><i class="fa fa-check"></i><b>5.2</b> Detecting SNP associations</a></li>
<li class="chapter" data-level="5.3" data-path="gwas-and-logistic-regression.html"><a href="gwas-and-logistic-regression.html"><i class="fa fa-check"></i><b>5.3</b> GWAS and logistic regression</a></li>
<li class="chapter" data-level="5.4" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html"><i class="fa fa-check"></i><b>5.4</b> Negative binomial and Poisson regression</a><ul>
<li class="chapter" data-level="5.4.1" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html#count-based-glms"><i class="fa fa-check"></i><b>5.4.1</b> Count-based GLMs</a></li>
<li class="chapter" data-level="5.4.2" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html#fitting-a-glm"><i class="fa fa-check"></i><b>5.4.2</b> Fitting a GLM</a></li>
<li class="chapter" data-level="5.4.3" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html#comparing-nested-models"><i class="fa fa-check"></i><b>5.4.3</b> Comparing nested models</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="negative-binomial-vs-poisson-glms.html"><a href="negative-binomial-vs-poisson-glms.html"><i class="fa fa-check"></i><b>5.5</b> Negative-Binomial vs Poisson GLMs</a></li>
<li class="chapter" data-level="5.6" data-path="further-understanding-the-model-optional.html"><a href="further-understanding-the-model-optional.html"><i class="fa fa-check"></i><b>5.6</b> Further understanding the model (<strong>OPTIONAL</strong>)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Essentials of Mathematics and Statistics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="detecting-snp-associations" class="section level2">
<h2><span class="header-section-number">5.2</span> Detecting SNP associations</h2>
<p>We have seen in lectures that we can do statistical tests for this type of contingency table using Chi Squared Tests. Let’s load example data set and and prepare</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="detecting-snp-associations.html#cb69-1"></a><span class="kw">library</span>(ggplot2) <span class="co"># for plots later</span></span>
<span id="cb69-2"><a href="detecting-snp-associations.html#cb69-2"></a><span class="kw">load</span>(<span class="st">&quot;gwas-cc-ex1.Rdata&quot;</span>)</span>
<span id="cb69-3"><a href="detecting-snp-associations.html#cb69-3"></a></span>
<span id="cb69-4"><a href="detecting-snp-associations.html#cb69-4"></a><span class="co"># how many individuals are there</span></span>
<span id="cb69-5"><a href="detecting-snp-associations.html#cb69-5"></a>n &lt;-<span class="st"> </span><span class="kw">length</span>(y)</span>
<span id="cb69-6"><a href="detecting-snp-associations.html#cb69-6"></a><span class="co"># How many SNPs do we have data for</span></span>
<span id="cb69-7"><a href="detecting-snp-associations.html#cb69-7"></a>p &lt;-<span class="st"> </span><span class="kw">nrow</span>(X)</span>
<span id="cb69-8"><a href="detecting-snp-associations.html#cb69-8"></a></span>
<span id="cb69-9"><a href="detecting-snp-associations.html#cb69-9"></a><span class="co"># samples that are controls are encoded as 0 in y</span></span>
<span id="cb69-10"><a href="detecting-snp-associations.html#cb69-10"></a>control &lt;-<span class="st"> </span><span class="kw">which</span>(y <span class="op">==</span><span class="st"> </span><span class="dv">0</span>)</span>
<span id="cb69-11"><a href="detecting-snp-associations.html#cb69-11"></a><span class="co"># disease cases are encoded as 1 in y</span></span>
<span id="cb69-12"><a href="detecting-snp-associations.html#cb69-12"></a>cases &lt;-<span class="st"> </span><span class="kw">which</span>(y <span class="op">==</span><span class="st"> </span><span class="dv">1</span>)</span></code></pre></div>
<p>Now we need to write a loop that scans through all, <span class="math inline">\(p\)</span>, SNPs:</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="detecting-snp-associations.html#cb70-1"></a><span class="co"># create a vector where p-values will be stored</span></span>
<span id="cb70-2"><a href="detecting-snp-associations.html#cb70-2"></a>p_vals &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, p)</span>
<span id="cb70-3"><a href="detecting-snp-associations.html#cb70-3"></a></span>
<span id="cb70-4"><a href="detecting-snp-associations.html#cb70-4"></a><span class="co"># Loop over SNPs</span></span>
<span id="cb70-5"><a href="detecting-snp-associations.html#cb70-5"></a><span class="cf">for</span> (i_p <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>p) {</span>
<span id="cb70-6"><a href="detecting-snp-associations.html#cb70-6"></a>    <span class="co"># 1. obtain genotype counts</span></span>
<span id="cb70-7"><a href="detecting-snp-associations.html#cb70-7"></a>    counts &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">ncol =</span> <span class="dv">3</span>)</span>
<span id="cb70-8"><a href="detecting-snp-associations.html#cb70-8"></a>    counts[<span class="dv">1</span>, ] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>),</span>
<span id="cb70-9"><a href="detecting-snp-associations.html#cb70-9"></a>                     <span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>),</span>
<span id="cb70-10"><a href="detecting-snp-associations.html#cb70-10"></a>                     <span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">2</span>))</span>
<span id="cb70-11"><a href="detecting-snp-associations.html#cb70-11"></a></span>
<span id="cb70-12"><a href="detecting-snp-associations.html#cb70-12"></a>    counts[<span class="dv">2</span>, ] &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>),</span>
<span id="cb70-13"><a href="detecting-snp-associations.html#cb70-13"></a>                     <span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>),</span>
<span id="cb70-14"><a href="detecting-snp-associations.html#cb70-14"></a>                     <span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">2</span>))</span>
<span id="cb70-15"><a href="detecting-snp-associations.html#cb70-15"></a></span>
<span id="cb70-16"><a href="detecting-snp-associations.html#cb70-16"></a>    <span class="co"># 2. expected probability of AA</span></span>
<span id="cb70-17"><a href="detecting-snp-associations.html#cb70-17"></a>    <span class="co"># (assuming no dependence on case/control status)</span></span>
<span id="cb70-18"><a href="detecting-snp-associations.html#cb70-18"></a>    expected_pr_AA &lt;-<span class="st"> </span><span class="kw">sum</span>(counts[, <span class="dv">1</span>]) <span class="op">/</span><span class="st"> </span>n</span>
<span id="cb70-19"><a href="detecting-snp-associations.html#cb70-19"></a>    <span class="co"># expected probability of Aa</span></span>
<span id="cb70-20"><a href="detecting-snp-associations.html#cb70-20"></a>    expected_pr_Aa &lt;-<span class="st"> </span><span class="kw">sum</span>(counts[, <span class="dv">2</span>]) <span class="op">/</span><span class="st"> </span>n</span>
<span id="cb70-21"><a href="detecting-snp-associations.html#cb70-21"></a>    <span class="co"># expected probability of aa</span></span>
<span id="cb70-22"><a href="detecting-snp-associations.html#cb70-22"></a>    expected_pr_aa &lt;-<span class="st"> </span><span class="kw">sum</span>(counts[, <span class="dv">3</span>]) <span class="op">/</span><span class="st"> </span>n</span>
<span id="cb70-23"><a href="detecting-snp-associations.html#cb70-23"></a></span>
<span id="cb70-24"><a href="detecting-snp-associations.html#cb70-24"></a>    expected_probs &lt;-<span class="st"> </span><span class="kw">c</span>(expected_pr_AA, expected_pr_Aa, expected_pr_aa )</span>
<span id="cb70-25"><a href="detecting-snp-associations.html#cb70-25"></a></span>
<span id="cb70-26"><a href="detecting-snp-associations.html#cb70-26"></a>    <span class="co"># 3. do my chi-squared test</span></span>
<span id="cb70-27"><a href="detecting-snp-associations.html#cb70-27"></a>    out &lt;-<span class="st"> </span><span class="kw">chisq.test</span>(counts, <span class="dt">p =</span> expected_probs)</span>
<span id="cb70-28"><a href="detecting-snp-associations.html#cb70-28"></a>    <span class="co"># extract p value of test and store</span></span>
<span id="cb70-29"><a href="detecting-snp-associations.html#cb70-29"></a>    p_vals[i_p] &lt;-<span class="st"> </span>out<span class="op">$</span>p.value</span>
<span id="cb70-30"><a href="detecting-snp-associations.html#cb70-30"></a>}</span></code></pre></div>
<p>We went through each SNP (rows in matrix <code>X</code>), extracted the counts of each genotype (marked <code>1.</code> in code) for cases and controls, then we compute expected probability (marked <code>2.</code> in code). Finally, we perform a chi-squared contingency table test comparing those observed counts to expected probabilities assuming that genotype is not related to disease status (marked <code>3.</code> in code).</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="detecting-snp-associations.html#cb71-1"></a>p_val_df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">p_val =</span> p_vals, <span class="dt">idx =</span> <span class="dv">1</span><span class="op">:</span>p)</span>
<span id="cb71-2"><a href="detecting-snp-associations.html#cb71-2"></a></span>
<span id="cb71-3"><a href="detecting-snp-associations.html#cb71-3"></a><span class="kw">ggplot</span>(p_val_df, <span class="kw">aes</span>(<span class="dt">x =</span> idx, <span class="dt">y =</span> <span class="op">-</span><span class="kw">log10</span>(p_val))) <span class="op">+</span></span>
<span id="cb71-4"><a href="detecting-snp-associations.html#cb71-4"></a><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="fl">2.5</span>, <span class="dt">col =</span> <span class="st">&quot;dodgerblue1&quot;</span>)</span></code></pre></div>
<p><img src="Practicals-module_1_files/figure-html/plot-1.png" width="960" /></p>
<p>This plot is knows as a <em>Manhattan</em> plot. One SNP (<code>i_p = 250</code>) will pop out as being highly associated with the disease process. Look at the genotype counts (or MAF) for this SNP in the cases and controls to see for yourself that there is large difference in the distribution of genotypes (or MAF).</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="detecting-snp-associations.html#cb72-1"></a>i_p &lt;-<span class="st"> </span><span class="dv">250</span></span>
<span id="cb72-2"><a href="detecting-snp-associations.html#cb72-2"></a>counts_v &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>), <span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>),</span>
<span id="cb72-3"><a href="detecting-snp-associations.html#cb72-3"></a>              <span class="kw">sum</span>(X[i_p, control] <span class="op">==</span><span class="st"> </span><span class="dv">2</span>), <span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">0</span>),</span>
<span id="cb72-4"><a href="detecting-snp-associations.html#cb72-4"></a>              <span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">1</span>), <span class="kw">sum</span>(X[i_p, cases] <span class="op">==</span><span class="st"> </span><span class="dv">2</span>))</span>
<span id="cb72-5"><a href="detecting-snp-associations.html#cb72-5"></a></span>
<span id="cb72-6"><a href="detecting-snp-associations.html#cb72-6"></a>snp_procs &lt;-<span class="st"> </span><span class="kw">data.frame</span>(counts_v, <span class="dt">type =</span> <span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;control&quot;</span>, <span class="st">&quot;cases&quot;</span>), <span class="dt">each =</span> <span class="dv">3</span>),</span>
<span id="cb72-7"><a href="detecting-snp-associations.html#cb72-7"></a>           <span class="dt">genotype =</span> <span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">&quot;AA&quot;</span>, <span class="st">&quot;Aa&quot;</span>, <span class="st">&quot;aa&quot;</span>), <span class="dv">2</span>))</span>
<span id="cb72-8"><a href="detecting-snp-associations.html#cb72-8"></a></span>
<span id="cb72-9"><a href="detecting-snp-associations.html#cb72-9"></a><span class="kw">ggplot</span>(snp_procs, <span class="kw">aes</span>(<span class="dt">x =</span> genotype, <span class="dt">y =</span> counts_v, <span class="dt">fill =</span> type)) <span class="op">+</span></span>
<span id="cb72-10"><a href="detecting-snp-associations.html#cb72-10"></a><span class="st">    </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">&quot;identity&quot;</span>, <span class="dt">position =</span> <span class="st">&quot;dodge&quot;</span>)</span></code></pre></div>
<p><img src="Practicals-module_1_files/figure-html/snp-1.png" width="768" /></p>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="data-2.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="gwas-and-logistic-regression.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="assets/gitbook-2.6.7/js/app.min.js"></script>
<script src="assets/gitbook-2.6.7/js/lunr.js"></script>
<script src="assets/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="assets/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["Practicals-module_1.pdf"],
"toc": {
"collapse": "subsection"
},
"toolbar": {
"position": "static"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
