<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5.6 Further understanding the model (OPTIONAL) | Essentials of Mathematics and Statistics</title>
  <meta name="description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
  <meta name="generator" content="bookdown 0.14 and GitBook 2.6.7" />

  <meta property="og:title" content="5.6 Further understanding the model (OPTIONAL) | Essentials of Mathematics and Statistics" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://anasrana.github.io/module1-practical_Bham/" />
  
  <meta property="og:description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
  <meta name="github-repo" content="anasrana/module1-practical_Bham" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5.6 Further understanding the model (OPTIONAL) | Essentials of Mathematics and Statistics" />
  
  <meta name="twitter:description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
  

<meta name="author" content="Anas A Rana" />


<meta name="date" content="2019-10-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="negative-binomial-vs-poisson-glms.html"/>

<script src="assets/jquery-2.2.3/jquery.min.js"></script>
<link href="assets/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Essentials of Mathematics and Statistics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="prerequisites.html"><a href="prerequisites.html"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="data-sets.html"><a href="data-sets.html"><i class="fa fa-check"></i><b>1.2</b> Data sets</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="practical-linear-regression.html"><a href="practical-linear-regression.html"><i class="fa fa-check"></i><b>2</b> Practical: Linear regression</a><ul>
<li class="chapter" data-level="2.1" data-path="data.html"><a href="data.html"><i class="fa fa-check"></i><b>2.1</b> Data</a></li>
<li class="chapter" data-level="2.2" data-path="simulating-data.html"><a href="simulating-data.html"><i class="fa fa-check"></i><b>2.2</b> Simulating data</a></li>
<li class="chapter" data-level="2.3" data-path="fitting-simple-linear-regression-model.html"><a href="fitting-simple-linear-regression-model.html"><i class="fa fa-check"></i><b>2.3</b> Fitting simple linear regression model</a><ul>
<li class="chapter" data-level="2.3.1" data-path="fitting-simple-linear-regression-model.html"><a href="fitting-simple-linear-regression-model.html#least-squared-estimation"><i class="fa fa-check"></i><b>2.3.1</b> Least squared estimation</a></li>
<li class="chapter" data-level="2.3.2" data-path="fitting-simple-linear-regression-model.html"><a href="fitting-simple-linear-regression-model.html#maximum-likelihood-estimation"><i class="fa fa-check"></i><b>2.3.2</b> Maximum likelihood estimation</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="effect-of-variance.html"><a href="effect-of-variance.html"><i class="fa fa-check"></i><b>2.4</b> Effect of variance</a></li>
<li class="chapter" data-level="2.5" data-path="exercise.html"><a href="exercise.html"><i class="fa fa-check"></i><b>2.5</b> Exercise</a><ul>
<li class="chapter" data-level="2.5.1" data-path="exercise.html"><a href="exercise.html#part-i"><i class="fa fa-check"></i><b>2.5.1</b> Part I</a></li>
<li class="chapter" data-level="2.5.2" data-path="exercise.html"><a href="exercise.html#part-ii"><i class="fa fa-check"></i><b>2.5.2</b> Part II</a></li>
<li class="chapter" data-level="2.5.3" data-path="exercise.html"><a href="exercise.html#part-iii"><i class="fa fa-check"></i><b>2.5.3</b> Part III</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="model-answers-linear-regression.html"><a href="model-answers-linear-regression.html"><i class="fa fa-check"></i>Model answers: Linear regression</a><ul>
<li class="chapter" data-level="2.6" data-path="exercise-i.html"><a href="exercise-i.html"><i class="fa fa-check"></i><b>2.6</b> Exercise I</a></li>
<li class="chapter" data-level="2.7" data-path="exercise-ii.html"><a href="exercise-ii.html"><i class="fa fa-check"></i><b>2.7</b> Exercise II</a><ul>
<li class="chapter" data-level="2.7.1" data-path="exercise-ii.html"><a href="exercise-ii.html#comparison-of-data"><i class="fa fa-check"></i><b>2.7.1</b> Comparison of data</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="exercise-ii-1.html"><a href="exercise-ii-1.html"><i class="fa fa-check"></i><b>2.8</b> Exercise II</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="practical-principal-component-analysis.html"><a href="practical-principal-component-analysis.html"><i class="fa fa-check"></i><b>3</b> Practical: Principal component analysis</a><ul>
<li class="chapter" data-level="3.1" data-path="data-1.html"><a href="data-1.html"><i class="fa fa-check"></i><b>3.1</b> Data</a></li>
<li class="chapter" data-level="3.2" data-path="introduction-1.html"><a href="introduction-1.html"><i class="fa fa-check"></i><b>3.2</b> Introduction</a></li>
<li class="chapter" data-level="3.3" data-path="exercise-i-1.html"><a href="exercise-i-1.html"><i class="fa fa-check"></i><b>3.3</b> Exercise I</a></li>
<li class="chapter" data-level="3.4" data-path="exercise-ii-2.html"><a href="exercise-ii-2.html"><i class="fa fa-check"></i><b>3.4</b> Exercise II</a></li>
<li class="chapter" data-level="3.5" data-path="exercise-iii.html"><a href="exercise-iii.html"><i class="fa fa-check"></i><b>3.5</b> Exercise III</a></li>
<li class="chapter" data-level="3.6" data-path="exercise-iv-single-cell-data.html"><a href="exercise-iv-single-cell-data.html"><i class="fa fa-check"></i><b>3.6</b> Exercise IV: Single cell data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="practical-multivariate-regression.html"><a href="practical-multivariate-regression.html"><i class="fa fa-check"></i><b>4</b> Practical: Multivariate regression</a></li>
<li class="chapter" data-level="5" data-path="practical-generalised-linear-models.html"><a href="practical-generalised-linear-models.html"><i class="fa fa-check"></i><b>5</b> Practical: Generalised linear models</a><ul>
<li class="chapter" data-level="5.1" data-path="data-2.html"><a href="data-2.html"><i class="fa fa-check"></i><b>5.1</b> Data</a></li>
<li class="chapter" data-level="5.2" data-path="detecting-snp-associations.html"><a href="detecting-snp-associations.html"><i class="fa fa-check"></i><b>5.2</b> Detecting SNP associations</a></li>
<li class="chapter" data-level="5.3" data-path="gwas-and-logistic-regression.html"><a href="gwas-and-logistic-regression.html"><i class="fa fa-check"></i><b>5.3</b> GWAS and logistic regression</a></li>
<li class="chapter" data-level="5.4" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html"><i class="fa fa-check"></i><b>5.4</b> Negative binomial and Poisson regression</a><ul>
<li class="chapter" data-level="5.4.1" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html#count-based-glms"><i class="fa fa-check"></i><b>5.4.1</b> Count-based GLMs</a></li>
<li class="chapter" data-level="5.4.2" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html#fitting-a-glm"><i class="fa fa-check"></i><b>5.4.2</b> Fitting a GLM</a></li>
<li class="chapter" data-level="5.4.3" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html#comparing-nested-models"><i class="fa fa-check"></i><b>5.4.3</b> Comparing nested models</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="negative-binomial-vs-poisson-glms.html"><a href="negative-binomial-vs-poisson-glms.html"><i class="fa fa-check"></i><b>5.5</b> Negative-Binomial vs Poisson GLMs</a></li>
<li class="chapter" data-level="5.6" data-path="further-understanding-the-model-optional.html"><a href="further-understanding-the-model-optional.html"><i class="fa fa-check"></i><b>5.6</b> Further understanding the model (<strong>OPTIONAL</strong>)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Essentials of Mathematics and Statistics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="further-understanding-the-model-optional" class="section level2">
<h2><span class="header-section-number">5.6</span> Further understanding the model (<strong>OPTIONAL</strong>)</h2>
<p>For assistance in further understanding the model, we can look at predicted counts for various levels of our predictors. Below we create new datasets with values of <code>immunoscore</code> and <code>tumor_subtype</code> and then use the predict command to calculate the predicted number of overexpressed proteins</p>
<p>First, we can look at predicted counts for each value of <code>tumor_subtype</code> while holding <code>immunoscore</code> at its mean. To do this, we create a new dataset with the combinations of <code>tumor_subtype</code> and <code>immunoscore</code> for which we would like to find predicted values, then use the <code>predict()</code> command.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="further-understanding-the-model-optional.html#cb88-1"></a>newdata_<span class="dv">1</span> &lt;-</span>
<span id="cb88-2"><a href="further-understanding-the-model-optional.html#cb88-2"></a><span class="kw">data.frame</span>(</span>
<span id="cb88-3"><a href="further-understanding-the-model-optional.html#cb88-3"></a>    <span class="dt">immunoscore =</span> <span class="kw">mean</span>(dat<span class="op">$</span>immunoscore),</span>
<span id="cb88-4"><a href="further-understanding-the-model-optional.html#cb88-4"></a>    <span class="dt">tumor_subtype =</span> <span class="kw">factor</span>(<span class="kw">c</span>(<span class="st">&quot;Complex&quot;</span>, <span class="st">&quot;Unstable&quot;</span>, <span class="st">&quot;Stable&quot;</span>), <span class="dt">labels =</span> <span class="kw">levels</span>(dat<span class="op">$</span>tumor_subtype)),</span>
<span id="cb88-5"><a href="further-understanding-the-model-optional.html#cb88-5"></a>    <span class="dt">gender=</span><span class="st">&quot;male&quot;</span>)</span>
<span id="cb88-6"><a href="further-understanding-the-model-optional.html#cb88-6"></a></span>
<span id="cb88-7"><a href="further-understanding-the-model-optional.html#cb88-7"></a>newdata_<span class="dv">2</span> &lt;-</span>
<span id="cb88-8"><a href="further-understanding-the-model-optional.html#cb88-8"></a><span class="kw">data.frame</span>(</span>
<span id="cb88-9"><a href="further-understanding-the-model-optional.html#cb88-9"></a>    <span class="dt">immunoscore =</span> <span class="kw">mean</span>(dat<span class="op">$</span>immunoscore),</span>
<span id="cb88-10"><a href="further-understanding-the-model-optional.html#cb88-10"></a>    <span class="dt">tumor_subtype =</span> <span class="kw">factor</span>(<span class="kw">c</span>(<span class="st">&quot;Complex&quot;</span>, <span class="st">&quot;Unstable&quot;</span>, <span class="st">&quot;Stable&quot;</span>), <span class="dt">labels =</span> <span class="kw">levels</span>(dat<span class="op">$</span>tumor_subtype)),</span>
<span id="cb88-11"><a href="further-understanding-the-model-optional.html#cb88-11"></a>    <span class="dt">gender=</span><span class="st">&quot;female&quot;</span>)</span>
<span id="cb88-12"><a href="further-understanding-the-model-optional.html#cb88-12"></a></span>
<span id="cb88-13"><a href="further-understanding-the-model-optional.html#cb88-13"></a>new_data &lt;-<span class="st"> </span><span class="kw">rbind</span>(newdata_<span class="dv">1</span>, newdata_<span class="dv">2</span>)</span>
<span id="cb88-14"><a href="further-understanding-the-model-optional.html#cb88-14"></a></span>
<span id="cb88-15"><a href="further-understanding-the-model-optional.html#cb88-15"></a>new_data<span class="op">$</span>phat &lt;-<span class="st"> </span><span class="kw">predict</span>(glm_<span class="dv">1</span>, new_data, <span class="dt">type =</span> <span class="st">&quot;response&quot;</span>)</span>
<span id="cb88-16"><a href="further-understanding-the-model-optional.html#cb88-16"></a></span>
<span id="cb88-17"><a href="further-understanding-the-model-optional.html#cb88-17"></a><span class="kw">print</span>(new_data)</span></code></pre></div>
<pre><code>##   immunoscore tumor_subtype gender      phat
## 1    48.26752       Complex   male  8.983829
## 2    48.26752        Stable   male  2.567187
## 3    48.26752      Unstable   male  5.876060
## 4    48.26752       Complex female 11.095193
## 5    48.26752        Stable female  3.170523
## 6    48.26752      Unstable female  7.257042</code></pre>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="further-understanding-the-model-optional.html#cb90-1"></a>newdata_<span class="dv">3</span> &lt;-</span>
<span id="cb90-2"><a href="further-understanding-the-model-optional.html#cb90-2"></a><span class="kw">data.frame</span>(</span>
<span id="cb90-3"><a href="further-understanding-the-model-optional.html#cb90-3"></a>    <span class="dt">immunoscore =</span> <span class="kw">rep</span>(<span class="kw">seq</span>(<span class="dt">from =</span> <span class="kw">min</span>(dat<span class="op">$</span>immunoscore), <span class="dt">to =</span> <span class="kw">max</span>(dat<span class="op">$</span>immunoscore), <span class="dt">length.out =</span> <span class="dv">100</span>), <span class="dv">3</span>),</span>
<span id="cb90-4"><a href="further-understanding-the-model-optional.html#cb90-4"></a>    <span class="dt">tumor_subtype =</span> <span class="kw">rep</span>(<span class="kw">factor</span>(<span class="kw">c</span>(<span class="st">&quot;Complex&quot;</span>, <span class="st">&quot;Unstable&quot;</span>, <span class="st">&quot;Stable&quot;</span>), <span class="dt">labels =</span> <span class="kw">levels</span>(dat<span class="op">$</span>tumor_subtype)), <span class="dt">each=</span><span class="dv">100</span>),</span>
<span id="cb90-5"><a href="further-understanding-the-model-optional.html#cb90-5"></a>    <span class="dt">gender=</span><span class="st">&quot;male&quot;</span>)</span>
<span id="cb90-6"><a href="further-understanding-the-model-optional.html#cb90-6"></a></span>
<span id="cb90-7"><a href="further-understanding-the-model-optional.html#cb90-7"></a>newdata_<span class="dv">4</span> &lt;-</span>
<span id="cb90-8"><a href="further-understanding-the-model-optional.html#cb90-8"></a><span class="kw">data.frame</span>(</span>
<span id="cb90-9"><a href="further-understanding-the-model-optional.html#cb90-9"></a>    <span class="dt">immunoscore =</span> <span class="kw">rep</span>(<span class="kw">seq</span>(<span class="dt">from =</span> <span class="kw">min</span>(dat<span class="op">$</span>immunoscore), <span class="dt">to =</span> <span class="kw">max</span>(dat<span class="op">$</span>immunoscore), <span class="dt">length.out =</span> <span class="dv">100</span>), <span class="dv">3</span>),</span>
<span id="cb90-10"><a href="further-understanding-the-model-optional.html#cb90-10"></a>    <span class="dt">tumor_subtype =</span> <span class="kw">rep</span>(<span class="kw">factor</span>(<span class="kw">c</span>(<span class="st">&quot;Complex&quot;</span>, <span class="st">&quot;Unstable&quot;</span>, <span class="st">&quot;Stable&quot;</span>), <span class="dt">labels =</span> <span class="kw">levels</span>(dat<span class="op">$</span>tumor_subtype)), <span class="dt">each=</span><span class="dv">100</span>),</span>
<span id="cb90-11"><a href="further-understanding-the-model-optional.html#cb90-11"></a>    <span class="dt">gender=</span><span class="st">&quot;female&quot;</span>)</span>
<span id="cb90-12"><a href="further-understanding-the-model-optional.html#cb90-12"></a></span>
<span id="cb90-13"><a href="further-understanding-the-model-optional.html#cb90-13"></a>new_data &lt;-<span class="st"> </span><span class="kw">rbind</span>(newdata_<span class="dv">3</span>, newdata_<span class="dv">4</span>)</span>
<span id="cb90-14"><a href="further-understanding-the-model-optional.html#cb90-14"></a></span>
<span id="cb90-15"><a href="further-understanding-the-model-optional.html#cb90-15"></a>new_data &lt;-<span class="st"> </span><span class="kw">cbind</span>(new_data, <span class="kw">predict</span>(glm_<span class="dv">1</span>, new_data, <span class="dt">type =</span> <span class="st">&quot;link&quot;</span>, <span class="dt">se.fit=</span><span class="ot">TRUE</span>))</span>
<span id="cb90-16"><a href="further-understanding-the-model-optional.html#cb90-16"></a></span>
<span id="cb90-17"><a href="further-understanding-the-model-optional.html#cb90-17"></a>new_data &lt;-<span class="st"> </span><span class="kw">within</span>(new_data, {</span>
<span id="cb90-18"><a href="further-understanding-the-model-optional.html#cb90-18"></a>  overexpressed_proteins &lt;-<span class="st"> </span><span class="kw">exp</span>(fit)</span>
<span id="cb90-19"><a href="further-understanding-the-model-optional.html#cb90-19"></a>  LL &lt;-<span class="st"> </span><span class="kw">exp</span>(fit <span class="op">-</span><span class="st"> </span><span class="fl">1.96</span> <span class="op">*</span><span class="st"> </span>se.fit)</span>
<span id="cb90-20"><a href="further-understanding-the-model-optional.html#cb90-20"></a>  UL &lt;-<span class="st"> </span><span class="kw">exp</span>(fit <span class="op">+</span><span class="st"> </span><span class="fl">1.96</span> <span class="op">*</span><span class="st"> </span>se.fit)</span>
<span id="cb90-21"><a href="further-understanding-the-model-optional.html#cb90-21"></a>})</span></code></pre></div>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="further-understanding-the-model-optional.html#cb91-1"></a><span class="kw">library</span>(ggplot2)</span>
<span id="cb91-2"><a href="further-understanding-the-model-optional.html#cb91-2"></a></span>
<span id="cb91-3"><a href="further-understanding-the-model-optional.html#cb91-3"></a><span class="kw">ggplot</span>(new_data, <span class="kw">aes</span>(immunoscore, overexpressed_proteins)) <span class="op">+</span></span>
<span id="cb91-4"><a href="further-understanding-the-model-optional.html#cb91-4"></a><span class="st">    </span><span class="kw">geom_ribbon</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> LL, <span class="dt">ymax =</span> UL, <span class="dt">fill =</span> tumor_subtype), <span class="dt">alpha =</span> <span class="fl">0.2</span>) <span class="op">+</span></span>
<span id="cb91-5"><a href="further-understanding-the-model-optional.html#cb91-5"></a><span class="st">    </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(<span class="dt">colour =</span> tumor_subtype), <span class="dt">size =</span> <span class="fl">1.5</span>) <span class="op">+</span></span>
<span id="cb91-6"><a href="further-understanding-the-model-optional.html#cb91-6"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Immunoscore&quot;</span>,</span>
<span id="cb91-7"><a href="further-understanding-the-model-optional.html#cb91-7"></a>         <span class="dt">y =</span> <span class="st">&quot;Overexpressed Proteins&quot;</span>) <span class="op">+</span></span>
<span id="cb91-8"><a href="further-understanding-the-model-optional.html#cb91-8"></a><span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>gender)</span></code></pre></div>
<p><img src="Practicals-module_1_files/figure-html/glm-plot-1.png" width="1152" /></p>

</div>
<!-- </div> -->
            </section>

          </div>
        </div>
      </div>
<a href="negative-binomial-vs-poisson-glms.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="assets/gitbook-2.6.7/js/app.min.js"></script>
<script src="assets/gitbook-2.6.7/js/lunr.js"></script>
<script src="assets/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="assets/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["Practicals-module_1.pdf"],
"toc": {
"collapse": "subsection"
},
"toolbar": {
"position": "static"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
