<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5.4 Negative binomial and Poisson regression | Essentials of Mathematics and Statistics</title>
  <meta name="description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
  <meta name="generator" content="bookdown 0.14 and GitBook 2.6.7" />

  <meta property="og:title" content="5.4 Negative binomial and Poisson regression | Essentials of Mathematics and Statistics" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://anasrana.github.io/module1-practical_Bham/" />
  
  <meta property="og:description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
  <meta name="github-repo" content="anasrana/module1-practical_Bham" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5.4 Negative binomial and Poisson regression | Essentials of Mathematics and Statistics" />
  
  <meta name="twitter:description" content="This contains practicals for the second week of module 1 (Essentials of Mathematics and Statistics)." />
  

<meta name="author" content="Anas A Rana" />


<meta name="date" content="2019-10-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="gwas-and-logistic-regression.html"/>
<link rel="next" href="negative-binomial-vs-poisson-glms.html"/>
<script src="assets/jquery-2.2.3/jquery.min.js"></script>
<link href="assets/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="assets/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Essentials of Mathematics and Statistics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="prerequisites.html"><a href="prerequisites.html"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="data-sets.html"><a href="data-sets.html"><i class="fa fa-check"></i><b>1.2</b> Data sets</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="practical-linear-regression.html"><a href="practical-linear-regression.html"><i class="fa fa-check"></i><b>2</b> Practical: Linear regression</a><ul>
<li class="chapter" data-level="2.1" data-path="data.html"><a href="data.html"><i class="fa fa-check"></i><b>2.1</b> Data</a></li>
<li class="chapter" data-level="2.2" data-path="simulating-data.html"><a href="simulating-data.html"><i class="fa fa-check"></i><b>2.2</b> Simulating data</a></li>
<li class="chapter" data-level="2.3" data-path="fitting-simple-linear-regression-model.html"><a href="fitting-simple-linear-regression-model.html"><i class="fa fa-check"></i><b>2.3</b> Fitting simple linear regression model</a><ul>
<li class="chapter" data-level="2.3.1" data-path="fitting-simple-linear-regression-model.html"><a href="fitting-simple-linear-regression-model.html#least-squared-estimation"><i class="fa fa-check"></i><b>2.3.1</b> Least squared estimation</a></li>
<li class="chapter" data-level="2.3.2" data-path="fitting-simple-linear-regression-model.html"><a href="fitting-simple-linear-regression-model.html#maximum-likelihood-estimation"><i class="fa fa-check"></i><b>2.3.2</b> Maximum likelihood estimation</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="effect-of-variance.html"><a href="effect-of-variance.html"><i class="fa fa-check"></i><b>2.4</b> Effect of variance</a></li>
<li class="chapter" data-level="2.5" data-path="exercise.html"><a href="exercise.html"><i class="fa fa-check"></i><b>2.5</b> Exercise</a><ul>
<li class="chapter" data-level="2.5.1" data-path="exercise.html"><a href="exercise.html#part-i"><i class="fa fa-check"></i><b>2.5.1</b> Part I</a></li>
<li class="chapter" data-level="2.5.2" data-path="exercise.html"><a href="exercise.html#part-ii"><i class="fa fa-check"></i><b>2.5.2</b> Part II</a></li>
<li class="chapter" data-level="2.5.3" data-path="exercise.html"><a href="exercise.html#part-iii"><i class="fa fa-check"></i><b>2.5.3</b> Part III</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="model-answers-linear-regression.html"><a href="model-answers-linear-regression.html"><i class="fa fa-check"></i>Model answers: Linear regression</a><ul>
<li class="chapter" data-level="2.6" data-path="exercise-i.html"><a href="exercise-i.html"><i class="fa fa-check"></i><b>2.6</b> Exercise I</a></li>
<li class="chapter" data-level="2.7" data-path="exercise-ii.html"><a href="exercise-ii.html"><i class="fa fa-check"></i><b>2.7</b> Exercise II</a><ul>
<li class="chapter" data-level="2.7.1" data-path="exercise-ii.html"><a href="exercise-ii.html#comparison-of-data"><i class="fa fa-check"></i><b>2.7.1</b> Comparison of data</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="exercise-ii-1.html"><a href="exercise-ii-1.html"><i class="fa fa-check"></i><b>2.8</b> Exercise II</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="practical-principal-component-analysis.html"><a href="practical-principal-component-analysis.html"><i class="fa fa-check"></i><b>3</b> Practical: Principal component analysis</a><ul>
<li class="chapter" data-level="3.1" data-path="data-1.html"><a href="data-1.html"><i class="fa fa-check"></i><b>3.1</b> Data</a></li>
<li class="chapter" data-level="3.2" data-path="introduction-1.html"><a href="introduction-1.html"><i class="fa fa-check"></i><b>3.2</b> Introduction</a></li>
<li class="chapter" data-level="3.3" data-path="exercise-i-1.html"><a href="exercise-i-1.html"><i class="fa fa-check"></i><b>3.3</b> Exercise I</a></li>
<li class="chapter" data-level="3.4" data-path="exercise-ii-2.html"><a href="exercise-ii-2.html"><i class="fa fa-check"></i><b>3.4</b> Exercise II</a></li>
<li class="chapter" data-level="3.5" data-path="exercise-iii.html"><a href="exercise-iii.html"><i class="fa fa-check"></i><b>3.5</b> Exercise III</a></li>
<li class="chapter" data-level="3.6" data-path="exercise-iv-single-cell-data.html"><a href="exercise-iv-single-cell-data.html"><i class="fa fa-check"></i><b>3.6</b> Exercise IV: Single cell data</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="practical-multivariate-regression.html"><a href="practical-multivariate-regression.html"><i class="fa fa-check"></i><b>4</b> Practical: Multivariate regression</a></li>
<li class="chapter" data-level="5" data-path="practical-generalised-linear-models.html"><a href="practical-generalised-linear-models.html"><i class="fa fa-check"></i><b>5</b> Practical: Generalised linear models</a><ul>
<li class="chapter" data-level="5.1" data-path="data-2.html"><a href="data-2.html"><i class="fa fa-check"></i><b>5.1</b> Data</a></li>
<li class="chapter" data-level="5.2" data-path="detecting-snp-associations.html"><a href="detecting-snp-associations.html"><i class="fa fa-check"></i><b>5.2</b> Detecting SNP associations</a></li>
<li class="chapter" data-level="5.3" data-path="gwas-and-logistic-regression.html"><a href="gwas-and-logistic-regression.html"><i class="fa fa-check"></i><b>5.3</b> GWAS and logistic regression</a></li>
<li class="chapter" data-level="5.4" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html"><i class="fa fa-check"></i><b>5.4</b> Negative binomial and Poisson regression</a><ul>
<li class="chapter" data-level="5.4.1" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html#count-based-glms"><i class="fa fa-check"></i><b>5.4.1</b> Count-based GLMs</a></li>
<li class="chapter" data-level="5.4.2" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html#fitting-a-glm"><i class="fa fa-check"></i><b>5.4.2</b> Fitting a GLM</a></li>
<li class="chapter" data-level="5.4.3" data-path="negative-binomial-and-poisson-regression.html"><a href="negative-binomial-and-poisson-regression.html#comparing-nested-models"><i class="fa fa-check"></i><b>5.4.3</b> Comparing nested models</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="negative-binomial-vs-poisson-glms.html"><a href="negative-binomial-vs-poisson-glms.html"><i class="fa fa-check"></i><b>5.5</b> Negative-Binomial vs Poisson GLMs</a></li>
<li class="chapter" data-level="5.6" data-path="further-understanding-the-model-optional.html"><a href="further-understanding-the-model-optional.html"><i class="fa fa-check"></i><b>5.6</b> Further understanding the model (<strong>OPTIONAL</strong>)</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Essentials of Mathematics and Statistics</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="negative-binomial-and-poisson-regression" class="section level2">
<h2><span class="header-section-number">5.4</span> Negative binomial and Poisson regression</h2>
<p>Molecular biologists study the behavior of protein expression in normal and cancerous tissues. The hypothesis is that the total number of over-expressed proteins depends on the histopathological-derived tumor subtype and an immune cell contexture measure.</p>
<p>You are provided with data on 314 tumours in the file <code>nb_data.Rdata</code>. The file contains one <code>data frame</code> with the following variables:</p>
<ul>
<li><strong><code>overexpressed_proteins</code></strong>: response variable of interest.</li>
<li><strong><code>immunoscore</code></strong>: gives a standardized measure of immune cell contexture.</li>
<li><strong><code>tumor_subtype</code></strong>: three-level nominal variable indicating the histopathological sub-type of the tumour. The three levels are Unstable, Stable, and Complex</li>
</ul>
<p>Let’s load some prerequisite R libraries and the data to produce some summary statistics (<em>install if required using <code>install.package()</code> command</em> ):</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="negative-binomial-and-poisson-regression.html#cb77-1"></a><span class="co"># required libraries</span></span>
<span id="cb77-2"><a href="negative-binomial-and-poisson-regression.html#cb77-2"></a><span class="kw">library</span>(MASS)</span>
<span id="cb77-3"><a href="negative-binomial-and-poisson-regression.html#cb77-3"></a><span class="kw">library</span>(foreign)</span>
<span id="cb77-4"><a href="negative-binomial-and-poisson-regression.html#cb77-4"></a></span>
<span id="cb77-5"><a href="negative-binomial-and-poisson-regression.html#cb77-5"></a></span>
<span id="cb77-6"><a href="negative-binomial-and-poisson-regression.html#cb77-6"></a><span class="kw">load</span>(<span class="st">&quot;nb_data.Rdata&quot;</span>)</span>
<span id="cb77-7"><a href="negative-binomial-and-poisson-regression.html#cb77-7"></a></span>
<span id="cb77-8"><a href="negative-binomial-and-poisson-regression.html#cb77-8"></a><span class="co"># print summary statistics to Console</span></span>
<span id="cb77-9"><a href="negative-binomial-and-poisson-regression.html#cb77-9"></a><span class="kw">summary</span>(dat)</span></code></pre></div>
<pre><code>##    sample_id      gender     immunoscore    overexpressed_proteins
##  1001   :  1   female:160   Min.   : 1.00   Min.   : 0.000        
##  1002   :  1   male  :154   1st Qu.:28.00   1st Qu.: 1.000        
##  1003   :  1                Median :48.00   Median : 4.000        
##  1004   :  1                Mean   :48.27   Mean   : 5.955        
##  1005   :  1                3rd Qu.:70.00   3rd Qu.: 8.000        
##  1006   :  1                Max.   :99.00   Max.   :35.000        
##  (Other):308                                                      
##   tumor_subtype
##  Complex : 40  
##  Unstable:167  
##  Stable  :107  
##                
##                
##                
## </code></pre>
<div id="count-based-glms" class="section level3">
<h3><span class="header-section-number">5.4.1</span> Count-based GLMs</h3>
<p>The <code>overexpressed_proteins</code> measurements are counts. This implies we should use a Poisson based GLM.</p>
<p>In Poisson regression models, the conditional variance is by definition equal to the conditional mean. This can be limiting.</p>
<p><em>Negative binomial regression</em> can be used for over-dispersed count data, that is when the conditional variance exceeds the conditional mean.</p>
<p>It can be considered as a generalization of Poisson regression since it has the same mean structure as Poisson regression but it has an extra parameter to model the over-dispersion. If the conditional distribution of the outcome variable is over-dispersed, the confidence intervals for the Poisson regression are likely to be narrower as compared to those from a Negative Binomial regression model.</p>
<p>In the following we will try both models to see which fits best.</p>
</div>
<div id="fitting-a-glm" class="section level3">
<h3><span class="header-section-number">5.4.2</span> Fitting a GLM</h3>
<p>Below we use the <code>glm.nb</code> function from the <code>MASS</code> package to estimate a negative binomial regression. The use of the function is similar to that of <code>lm</code> for linear models but with the additional requirement of a link function. As count data is always positive, a log link function is useful here.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="negative-binomial-and-poisson-regression.html#cb79-1"></a>glm_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">glm.nb</span>(overexpressed_proteins <span class="op">~</span><span class="st"> </span>immunoscore <span class="op">+</span><span class="st"> </span>tumor_subtype <span class="op">+</span><span class="st"> </span>gender, <span class="dt">data =</span> dat, <span class="dt">link=</span>log)</span>
<span id="cb79-2"><a href="negative-binomial-and-poisson-regression.html#cb79-2"></a></span>
<span id="cb79-3"><a href="negative-binomial-and-poisson-regression.html#cb79-3"></a><span class="co"># print summary statistics of glm.nb output object to Console</span></span>
<span id="cb79-4"><a href="negative-binomial-and-poisson-regression.html#cb79-4"></a><span class="kw">summary</span>(glm_<span class="dv">1</span>)</span></code></pre></div>
<pre><code>## 
## Call:
## glm.nb(formula = overexpressed_proteins ~ immunoscore + tumor_subtype + 
##     gender, data = dat, link = log, init.theta = 1.047288915)
## 
## Deviance Residuals: 
##     Min       1Q   Median       3Q      Max  
## -2.1567  -1.0761  -0.3810   0.2856   2.7235  
## 
## Coefficients:
##                        Estimate Std. Error z value Pr(&gt;|z|)    
## (Intercept)            2.707484   0.204275  13.254  &lt; 2e-16 ***
## immunoscore           -0.006236   0.002492  -2.502   0.0124 *  
## tumor_subtypeUnstable -0.424540   0.181725  -2.336   0.0195 *  
## tumor_subtypeStable   -1.252615   0.199699  -6.273 3.55e-10 ***
## gendermale            -0.211086   0.121989  -1.730   0.0836 .  
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## (Dispersion parameter for Negative Binomial(1.0473) family taken to be 1)
## 
##     Null deviance: 431.67  on 313  degrees of freedom
## Residual deviance: 358.87  on 309  degrees of freedom
## AIC: 1740.3
## 
## Number of Fisher Scoring iterations: 1
## 
## 
##               Theta:  1.047 
##           Std. Err.:  0.108 
## 
##  2 x log-likelihood:  -1728.307</code></pre>
<p>R first displays the call and the deviance residuals. Next, we see the regression coefficients for each of the variables, along with standard errors, z-scores, and p-values. The variable <code>immunoscore</code> has a coefficient of -0.006, which is statistically significant at the 5% level (<code>Pr(&gt;|z|) = 0.0124*</code>). This means that for each one-unit increase in <code>immunoscore</code>, the expected log count of the number of <code>overexpressed_proteins</code> decreases by 0.006.</p>
<p>The indicator variable shown as <code>tumor_subtypeUnstable</code> is the expected difference in log count between group Unstable and the reference group (<code>tumor_subtype=Complex</code>). The expected log count for the Unstable type is approximately 0.4 lower than the expected log count for the Complex type.</p>
<p>The indicator variable for Stable type is the expected difference in log count between the Stable type and the reference Complex group. The expected log count for Stable is approximately 1.2 lower than the expected log count for the Complex type.</p>
</div>
<div id="comparing-nested-models" class="section level3">
<h3><span class="header-section-number">5.4.3</span> Comparing nested models</h3>
<p>To determine if <code>tumor_subtype</code> itself, overall, is statistically significant, we can compare a model with and without <code>tumor_subtype</code>. The reason it is important to fit separate models is that, unless we do, the overdispersion parameter is held constant and it would not be a fair comparison.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="negative-binomial-and-poisson-regression.html#cb81-1"></a>glm_<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw">glm.nb</span>(overexpressed_proteins <span class="op">~</span><span class="st"> </span>immunoscore <span class="op">+</span><span class="st"> </span>gender, <span class="dt">data =</span> dat, <span class="dt">link =</span> log)</span></code></pre></div>
<p>We use the <code>anova</code> function to compare models using a likelihood ratio test (LRT):</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="negative-binomial-and-poisson-regression.html#cb82-1"></a><span class="kw">anova</span>(glm_<span class="dv">1</span>, glm_<span class="dv">2</span>, <span class="dt">test =</span> <span class="st">&quot;LRT&quot;</span>)</span></code></pre></div>
<pre><code>## Warning in anova.negbin(glm_1, glm_2, test = &quot;LRT&quot;): only Chi-squared LR
## tests are implemented</code></pre>
<pre><code>## Likelihood ratio tests of Negative Binomial Models
## 
## Response: overexpressed_proteins
##                                  Model     theta Resid. df    2 x log-lik.
## 1                 immunoscore + gender 0.8705939       311       -1772.074
## 2 immunoscore + tumor_subtype + gender 1.0472889       309       -1728.307
##     Test    df LR stat.      Pr(Chi)
## 1                                   
## 2 1 vs 2     2 43.76737 3.133546e-10</code></pre>
<p>The two degree-of-freedom chi-square test indicates that <code>tumor_subtype</code> is a statistically significant predictor of <code>overexpressed_proteins</code> (<code>Pr(Chi) = 3.133546e-10</code>).</p>
<p>The <code>anova</code> function performs a form of <em>LRT</em>. It computes the likelihood of the data under the two models being compared and then uses the ration of these likelihood values as a test statistic.</p>
<p>Theory tells us that, for large samples sizes, the (2x) log likelihood ratio has a chi-squared distribution with degrees of freedom equal to the difference in the number of free parameters between the two models being compared. The LRT only applies to <em>nested models</em>, i.e. a pair of models where one is a less complex subset of the other.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="gwas-and-logistic-regression.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="negative-binomial-vs-poisson-glms.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="assets/gitbook-2.6.7/js/app.min.js"></script>
<script src="assets/gitbook-2.6.7/js/lunr.js"></script>
<script src="assets/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="assets/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="assets/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["Practicals-module_1.pdf"],
"toc": {
"collapse": "subsection"
},
"toolbar": {
"position": "static"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
